### PR TITLE
Support text wrapping

### LINKED ISSUE #16: Support text wrapping
Hello, I get a box collapse when I want to print a string longer than the window width. Is it a normal behavior?

### Screenshot:

image

### Steps to reproduce:

```js
const boxen = require('boxen');
console.log(boxen('unicornddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd', {padding: 1}));
```

### PR DESCRIPTION
**Pull request by aichholzer**
Fixes #16

### CODE REVIEW

**Comment by aichholzer** [comment]
I reckon this problem is related to the same as discussed here: https://github.com/sindresorhus/boxen/pull/14 where the child process has no knowledge of the actual width.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "boxen",
-  "version": "1.1.0",
+  "version": "1.2.0",
```
Comment (SamVerschueren) [inline package.json]: Don't do this. It happens automatically when we publish.

**Comment by SamVerschueren** [comment]
I wonder, can't we use [wrap-ansi](https://github.com/chalk/wrap-ansi) instead? It's battle tested way more than a custom implementation.

**Comment by aichholzer** [comment]
@SamVerschueren feel free to reject this pull request if you prefer a `wrap-ansi` implementation. The proposed solution here is pretty straightforward and the logic involved is not overly complicated.

**Comment by SamVerschueren** [comment]
I'm not saying the PR or the implementation is bad, I'm just proposing alternatives. Thinking about what cases might be missed with the current implementation. For instance, it doesn't take color codes (ANSI escape codes) into account when wrapping. So with that in mind, I was just proposing `wrap-ansi` as an alternative which might cover more use cases and might be easier to use and maintain.

**Comment by sindresorhus** [comment]
@aichholzer Why did you close this?

**Comment by aichholzer** [comment]
@sindresorhus -It seemed like it was going to turn into a stale PR and I am all for cleaning up.

**Comment by sindresorhus** [comment]
I just haven't had time to review yet because of summer and lack of keyboard time. Will try to review asap.

File: index.js [hunk 2]
```diff
@@ -93,9 +94,14 @@ module.exports = (text, opts) => {
 	};
 
 	const colorizeContent = x => opts.backgroundColor ? chalk[opts.backgroundColor](x) : x;
+	const columns = termSize().columns;
 
 	text = ansiAlign(text, {align: opts.align});
 
+	if (text.length > columns) {
```
Comment (sindresorhus) [inline index.js]: `text.length` is not an accurate measure. Some characters can have `.length` of 3, but take up 2 columns. Better to just let `wrap-ansi` handle everything.

**Comment by sindresorhus** [comment]
Ping :)

**Comment by aichholzer** [comment]
Pong :)
@sindresorhus -Not sure what you mean. AFAIK, `wrap-ansi` does not provide an option to do a wrap at a certain column length, or perhaps I am just missing something here.

File: index.js (hunk 2 above)
Comment (aichholzer) [reply index.js]: @sindresorhus -Not sure what you mean. AFAIK, `wrap-ansi` does not provide an option to do a wrap at a certain column length (or to automatically wrap at terminal width), or perhaps I am just missing something here.

Comment (SamVerschueren) [reply index.js]: https://github.com/chalk/wrap-ansi/blob/master/readme.md#wrapansiinput-columns-options

Second argument

Comment (aichholzer) [reply index.js]: @SamVerschueren -Aware of that. My mistake in the previous comment. I did not explain myself well.

```
const columns = termSize().columns;
text = ansiAlign(text, {align: opts.align});

if (text.length > columns) {
  text = wrapAnsi(text, columns);
}
```

It is already being done like that. I was wondering about the `text.length` and `wrap-ansi` format;

```
text = ansiAlign(text, {align: 'center'});
text = wrapAnsi(text, columns);
```

`wrapAnsi` does not respect the `center` format, tests fail.

Comment (SamVerschueren) [reply index.js]: Can you reverse the calls?

```js
text = wrapAnsi(text, columns);
text = ansiAlign(text, {align: 'center'});
```

**Comment by aichholzer** [comment]
@SamVerschueren -Now that really made me look/feel/seem like a complete noob, didn't it?
Damn, today (this week) has not been a good one.

File: package.json [hunk 3]
```diff
@@ -33,15 +33,16 @@
...
+    "string-width": "^2.1.1",
     "term-size": "^1.2.0",
-    "widest-line": "^1.0.0"
+    "widest-line": "^1.0.0",
+    "wrap-ansi": "^3.0.1"
   },
   "devDependencies": {
-    "ava": "*",
-    "nyc": "^11.0.3",
+    "ava": "^0.22.0",
```
Comment (SamVerschueren) [inline package.json]: Revert it back to `*`

File: package.json [hunk 4]
```diff
@@ -33,15 +33,16 @@
...
+    "nyc": "^11.2.1",
```
Comment (SamVerschueren) [inline package.json]: Revert it back. Semver handles this.

File: package.json [hunk 5]
```diff
@@ -33,15 +33,16 @@
   "dependencies": {
     "ansi-align": "^2.0.0",
     "camelcase": "^4.0.0",
-    "chalk": "^2.0.1",
+    "chalk": "^2.1.0",
```
Comment (SamVerschueren) [inline package.json]: Revert it back. Semver handles this.

File: package.json [hunk 6]
```diff
@@ -33,15 +33,16 @@
...
     "cli-boxes": "^1.0.0",
-    "string-width": "^2.0.0",
+    "string-width": "^2.1.1",
```
Comment (SamVerschueren) [inline package.json]: Revert it back. Semver handles this.

File: package.json (hunk 5 above)
Comment (aichholzer) [reply package.json]: This came from upgrading the packages. Not intentionally.

**Comment by aichholzer** [comment]
@SamVerschueren @sindresorhus -Als je Nederlands spreekt; misschien zou het volgende keer beter zijn, voor mij tenminste!

**Comment by SamVerschueren** [comment]
> Als je Nederlands spreekt; misschien zou het volgende keer beter zijn, voor mij tenminste!

Ik spreek Nederlands, Sindre niet. Waarom zou het voor u gemakkelijker zijn? De common language binnen IT is sowieso Engels. Ook voor anderen die deze PR bekijken is het gemakkelijker als alles ook voor hen begrijpelijk is.

So let's just continue in english :).

File: package.json [hunk 7]
```diff
@@ -30,18 +30,19 @@
...
-    "string-width": "^2.0.0",
-    "term-size": "^1.2.0",
-    "widest-line": "^1.0.0"
-  },
-  "devDependencies": {
-    "ava": "*",
-    "nyc": "^11.0.3",
-    "xo": "*"
-  }
+	"dependencies": {
```
Comment (SamVerschueren) [inline package.json]: tabs -> spaces

**Comment by aichholzer** [comment]
@SamVerschueren -The comment in dutch was just meant to be funny, implying that I have certainly made myself look like a complete noob with it and hoping that next time I do a PR it would go better. For the rest; yes, english it is.

**Pull request closed without merging**